### ISSUE TITLE
[Question] Effects on EntityFramework Migrations

### ISSUE THREAD

**Issue by gentledepp** [Mimetis/Dotmim.Sync#30]
Hi
Yesterday, I was working on a software based on SyncWinrt, and I remembered, that it has a quite significant issue:

It requires us to run SQL scripts at setup time, so the synchronization can actually work.
As we are using EntityFramework Code First, we had to include these scripts in the EF migrations.

That, however, had a quite negative impact: We could not use [Effort  ](https://github.com/tamasflamich/effort) - an in-memory provider for EF - anymore.
This significantly slowed down our test performance.

So my question:
What is the best way to use Dotmim.Sync in combination with EF Migrations? 
Did you think about this? :-)

**Comment by Mimetis** [comment]
I think the best solution is to provide some usefuls methods to be able to provisioning and deprovisionning the database, from the Dotmim.Sync framework.   
So we can maybe do a description like this:
* Start test
* Create the test database with code first
* Make tests without sync
* Provision the database
* Make tests with sync
* Unprovision database
* Delete test database

I don't manage the schema changes for a current database.  
Actually the only way to work with schémas could be Something like:
* Create a database
* Activate the Dotmim.Sync sync (create triggers / tables/ stored procedure)

Ok, you want to add / delete columns on tables:
* Unprovision server database **without** loosing metadatas (delete triggers and stored procédures, lets the tracking tables alive)
* Edit the schema (with EF Migration, or directly with scripts)
* Reprovision (recreate triggers and stored procedure)

On the client side, it's pretty the same method:
* Unprovision client database, again without loosing metadatas
* Edit the schema
* Reprovision

Is that make sense for you ?

**Comment by Mimetis** [comment]
hey @gentledepp 

I m currently working on a solution, to be able to add some migrations stuff in your database evolution.

It's not actually available, but it will be in a maybe a couple of weeks, depending on all the tests I have to make to validate it !

Here is how it could works:

First of all, you have a straightforward sync process on 2 tables, Something like this:

``` cs
    public async Task SynchronizeAsync()
    {
        SqlSyncProvider serverProvider = new SqlSyncProvider(GetDatabaseConnectionString("Northwind"));
        SqlSyncProvider client1Provider = new SqlSyncProvider(GetDatabaseConnectionString("NW1"));

        SyncConfiguration configuration = new SyncConfiguration(new string[] {
        "Customers", "Region"});

        SyncAgent agent1 = new SyncAgent(client1Provider, serverProvider, configuration);

        var context = await agent1.SynchronizeAsync();

        // do stuff ..
    }
```
Now imagine you want to add new column, called `Comments` on the `Customers` table.   
Here is how it will work:

* Create the providers (server and client in my sample)
* Delete the stored procédures and the triggers
* We don't want to loose the `Customer` tracking table, so we don't delete it. Here we have to be sure the primary keys from `Customer` is still the same (So don't touch primary keys or this technic won't work)
* Edit the client and server schema (you can use here the EF migration)
* Re apply the triggers and the stored procédures. They will be re-generated with the new column !
* Re launch a new sync process (don't use the same `SyncAgent` as before, recreate a new one).

``` cs
private static async Task AlterSchemasAsync()
{
    SqlSyncProvider serverProvider = new SqlSyncProvider(GetDatabaseConnectionString("Northwind"));
    SqlSyncProvider clientProvider = new SqlSyncProvider(GetDatabaseConnectionString("NW1"));

    // tables to edit
    var tables = new string[] { "Customers" };

    // delete triggers and sp
    await serverProvider.DropSyncTablesAsync(tables, SyncProvision.StoredProcedures | SyncProvision.Triggers);
    await clientProvider.DropSyncTablesAsync(tables, SyncProvision.StoredProcedures | SyncProvision.Triggers);

    // use whatever you want to edit your schema
    // add column on server
    using (SqlConnection cs = serverProvider.CreateConnection() as SqlConnection)
    {
        cs.Open();
        SqlCommand cmd = new SqlCommand("ALTER TABLE dbo.Customers ADD Comments nvarchar(50) NULL", cs);
        cmd.ExecuteNonQuery();
        cs.Close();
    }
    // add column on client
    using (SqlConnection cs = clientProvider.CreateConnection() as SqlConnection)
    {
        cs.Open();
        SqlCommand cmd = new SqlCommand("ALTER TABLE dbo.Customers ADD Comments nvarchar(50) NULL", cs);
        cmd.ExecuteNonQuery();
        cs.Close();
    }

    // re apply server conf
    await serverProvider.CreateSyncTablesAsync(tables, SyncProvision.StoredProcedures | SyncProvision.Triggers);
    await clientProvider.CreateSyncTablesAsync(tables, SyncProvision.StoredProcedures | SyncProvision.Triggers);

  // sync !
  await this.SynchronizeAsync();
}
```

Is that make sense for you .? Is it Something you need ?

Seb

**Comment by gentledepp** [comment]
Hi!
This sounds nice. But I have some thoughts for you:

1.  I would very much prefer to keep all migration logic in the migrations which are autogenerated by entity framework.
So basically, I'd like to add the dotmim.sync code used to do the migration stuff into my code first migration class (and not the other way around as you just sketched it)
So it would be great if there were some System.Data.Entity.Migrations.Model.MigrationOperation s for this scenario (this is the class of EF 6.0 - I don't know how it is called in EF Core)
As an example, please have a look at the **Figure 6 A Custom Migration Operation" [here](Figure 6 A Custom Migration Operation for Creating a Database View)

2. What happens, when this migration (adding another column) to the clients, which have not (yet) upgraded their schema? Won't this break the synchronization process?
In my opinion, this can be solved two ways:

- There must be a way to define a version of this sync contract.
If the version of the client differs, it must re-sync all data (and not only partials)

- Or you need to create a new named sync scope and run the old and new versions (v1 and v2) in parallel. So clients can run different versions.
However, when a client updates, it must realize, that its current sync contract version is v1 and it requires v2 and then start a full sync.

Does that make sense to you?

**Comment by Mimetis** [comment]
* 1 The project is .Net Standard 2.0 with no dependencies on EF (6 or Core) So I can not provide a code-migration-like feature.
* 2 Since you'are are updating the client, I assume that you'are updating your software as well. You can so embed the sync code migration. We can't use different versions when you are upgrading a db, since columns won't be the same between one version and an other one.

I will push as soon as possible a new version, and I will write a sample doc about what can be a migration code.

**Comment by Mimetis** [comment]
Take a look on the documentation on the subject here : [https://mimetis.github.io/Dotmim.Sync/Provision](https://mimetis.github.io/Dotmim.Sync/Provision)

**Comment by gentledepp** [comment]
Hi! I am sorry for answering sooo late. I did not mark your e-mail as "not read" and simply forgot about it. Sorry about that.

As for your proposals - I think this is pretty much the way to go.
I completely forgot that Dotmim.Sync is based on .NET Standard. So what I could think of is putting all the Dotmim.Sync provisioning code into a EF Code First migration. (Because my EF migration can know about Dotmim.Sync I guess)

As for the schema change approach: I don't think this will work.
The reason is, that you cannot make sure that all clients out in the wild will update their schema at the same time.
You will have clients running different versions of the schema - this is inevitable.
Now let us consider the fact, that you add another `Comments`  column to the `Customers` table on the server. Then this column **must not** be required. 
The reason is, that there might be a client A out there, that has changes in that table (e.g. a user added another customer) and needs to push those back to the server. This would be impossible, if there `Comments` column was not nullable.

What I could think of is:
In the sync process, the client and server compare their schema version (I will call it "sync scope version" for now).
If they are the same, syncing just continues.
If the client has a lower version
   * but the changes are downwardly compatible (e.g. a column was deleted on the server and can further be ignored on the client), then the client just continues syncing
   * if the changes are not downwardly compatible (e.g. a column was added on the server), then the client needs to update its schema (either by running a migration script or telling the user to update the application itself!) and re-sync the entire scope.

In the latter case, I do not know what I'd do with local changes on the client, that need to be pushed to the server...  

Also note: For this to work, the server must always have the same or higher scope version than the client.

Did I make this problem clear? (I hope so :-|)
